src/picosvg/svg.py

```diff
@@ -1021,106 +1116,74 @@ def resolve_nested_svgs(self, inplace=False):
+        # split translation from rest of the transform and apply to gradient coords
+        translate, affine_prime = affine.decompose_translation()
+        if translate.round(_GRADIENT_TRANSFORM_NDIGITS) != Affine2D.identity():
+            for x_attr, y_attr in _GRADIENT_COORDS[strip_ns(el.tag)]:
+                x = getattr(gradient, x_attr)
+                y = getattr(gradient, y_attr)
+                x_prime, y_prime = translate.map_point((x, y))
+                setattr(gradient, x_attr, round(x_prime, _GRADIENT_TRANSFORM_NDIGITS))
+                setattr(gradient, y_attr, round(y_prime, _GRADIENT_TRANSFORM_NDIGITS))
 
-    def _apply_gradient_translation(self, inplace=False):
-        if not inplace:
-            svg = SVG(copy.deepcopy(self.svg_root))
-            svg._apply_gradient_translation(inplace=True)
-            return svg
-
-        for el in self._select_gradients():
-            gradient = _GRADIENT_CLASSES[strip_ns(el.tag)].from_element(
-                el, self.view_box()
-            )
-            affine = gradient.gradientTransform
-
-            # split translation from rest of the transform and apply to gradient coords
-            translate, affine_prime = affine.decompose_translation()
-            if translate.round(_GRADIENT_TRANSFORM_NDIGITS) != Affine2D.identity():
-                for x_attr, y_attr in _GRADIENT_COORDS[strip_ns(el.tag)]:
-                    x = getattr(gradient, x_attr)
-                    y = getattr(gradient, y_attr)
-                    x_prime, y_prime = translate.map_point((x, y))
-                    setattr(
-                        gradient, x_attr, round(x_prime, _GRADIENT_TRANSFORM_NDIGITS)
-                    )
-                    setattr(
-                        gradient, y_attr, round(y_prime, _GRADIENT_TRANSFORM_NDIGITS)
-                    )
+        gradient.gradientTransform = affine_prime.round(_GRADIENT_TRANSFORM_NDIGITS)
 
-            gradient.gradientTransform = affine_prime.round(_GRADIENT_TRANSFORM_NDIGITS)
+        el.attrib.clear()
+        el.attrib.update(to_element(gradient).attrib)
 
-            el.attrib.clear()
-            el.attrib.update(to_element(gradient).attrib)
-
-    def _resolve_gradient_templates(self, inplace=False):
+    def _apply_gradient_template(self, gradient: etree.Element):
         # Gradients can have an 'href' attribute that specifies another gradient as
         # a template, inheriting its attributes and/or stops when not already defined:
         # https://www.w3.org/TR/SVG/pservers.html#PaintServerTemplates
-        if not inplace:
-            svg = SVG(copy.deepcopy(self.svg_root))
-            svg._resolve_gradient_templates(inplace=True)
-            return svg
+
+        assert _is_gradient(gradient)
 
         href_attr = _xlink_href_attr_name()
+        if href_attr not in gradient.attrib:
+            return  # nop
+
         el_by_id = {el.attrib["id"]: el for el in self.xpath(".//svg:*[@id]")}
```

Comment:
here we are creating this `el_by_id` dictionary (xpath-ing the whole document for element ids) for each gradient, every time we call the new `_apply_gradient_template` method; before this was done only once for all gradients, in `_resolve_gradient_templates`, and use it to quickly look up elements by ids.
I suggest that you either optionally have an `el_by_id` parameter that caller of `_apply_gradient_template` can pass down, or you remove it altogether and do xpath for each gradient to search the element by the ref id

Reply:
I like option 2, I'll change to that I think.